Improve st.help (and st.write's usage of st.help!)

sfc-gh-tteixeira opened the pull request:
## 📚 Context

As part of my hackathon project, I'm planning on submitting some fully-ready PRs (not prototypes!) that hopefully can be merged without a lot of work from the eng team. This is one of them! If it's too much work to merge, or the feature is deemed a bad idea, I'm happy to drop the whole thing.

What this PR does:
* Improves st.help() in the following ways:
   * Shows the actual variable or expression you passed to st.help. For example st.help(foo[0]["bar"]) prints foo[0]["bar"] in the header of the displayed block, so you know what that block is referring to.
   * Shows the type is a cleaner, more readable way
   * Shows the docstring in a scrolling box, to save space
   * Shows the members of the thing you're asking for help on
      * ...along with their current values or docstrings (depending on whether it's a function or not)
* Improves st.write() in the following ways:
   * If you pass a class into st.write() you now get st.help(the_class) 
   * If you pass an object that doesn't translate well into a string (i.e. things that would normally look like `<foo object at 0x1234abcd>`), calls st.help(the_object).

This makes st.help, st.write, and magic great for quickly inspecting objects, classes, and what-have-yous!  More info in the [spec](https://www.notion.so/streamlit/Mini-spec-for-st-help-142c869006364bca8114f73e8ff654a7).

- What kind of change does this PR introduce?

  - [ ] Bugfix
  - [x] Feature
  - [ ] Refactoring
  - [ ] Other, please describe:

## 🧠 Description of Changes

- _Add bullet points summarizing your changes here_

  - [ ] This is a breaking API change
  - [x] This is a visible (user-facing) change

## 😬 Before

This is what you used to get for this code:
```python

class Animal:
    pass

class Dog(Animal):
    """A dog object.

    These kinds of classes are very useful in day-to-day programming,
    judging by my "Intro to OOP" book.
    """

    def __init__(self, color, breed):
        self.color = color
        self.breed = breed

    def bark(self, reason):
        """Make a pleasant sound."""
        pass

fido = Dog("brown", "poodle")


# Show help for an object instance.
st.help(fido)

# Show help for a class.
st.help(Dog)

# Show help for a module (useful to avoid consulting online docs!)
re
```

---
<img width="739" alt="Screen Shot 2022-12-14 at 20 42 57" src="https://user-images.githubusercontent.com/103002884/207774407-55b49911-7ea8-477c-a109-3f901151d151.png">
(etc)

---
## 😎 After
<img width="739" alt="Screen Shot 2022-12-14 at 20 43 01" src="https://user-images.githubusercontent.com/103002884/207774416-9dc31748-c686-451a-9abc-07f387e5f975.png">
(etc)


## 🧪 Testing Done

- [x] Screenshots included
- [x] Added/Updated unit tests
- [ ] Added/Updated e2e tests - **Still need to do this! Will do before merge.**

## 🌐 References

_Does this depend on other work, documents, or tickets?_

- **Issue**: none

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

sfc-gh-tteixeira committed:
Improve st.help and use it more in st.write.

sfc-gh-tteixeira committed:
Add more tests

sfc-gh-tteixeira committed:
Move imports into tests, so only the specific tests fail if the imports fail.

sfc-gh-tteixeira committed:
If st.help(obj_instance), sometimes show value for Class.

sfc-gh-tteixeira committed:
Support 3.7

jrieke commented:
Had a rough look through it and generally LGTM! Let's make a demo app in the feature-demos repo before we merge, so we can have a look what it looks like in live, and then I'll give my final LGTM.

sfc-gh-tteixeira committed:
Improve behavior inside exec() blocks.
(i.e. it works but doesn't show variable name)

sfc-gh-tteixeira committed:
Show warning when e2e/scripts/st_help.py is run within an exec() block
(this typically only happens in PR previews)

jrieke commented:
@sfc-gh-tteixeira What happens if I do `st.write("This is Fido:", fido)` in your example? Does it show 1) `This is Fido: <__main__.Dog object at ...>` or does it show 2) `This is Fido:` and then the help box below? I think in this case we should show (1) since I assume some people might do something like this to quickly check which value an object has. And if show the box instead, it kinda distorts their app. No super strong opinion though!

sfc-gh-tteixeira committed:
Merge remote-tracking branch 'origin/develop' into st-help

sfc-gh-tteixeira commented:
> Let's make a demo app in the feature-demos repo before we merge, so we can have a look what it looks like in live, and then I'll give my final LGTM.

Turns out the feature-demos app doesn't work very well with this. That's beacuse part of the new `st.help` code won't work inside an `exec` block (because of [this](https://stackoverflow.com/a/12072941)), i.e. the part that finds out variable names (in the red circle, in the screenshot below).

So I'm adding a warning to e2e/scripts/st_help.py to explain this, so people don't get confused when they run feature-demos.

<img width="719" alt="207774416-9dc31748-c686-451a-9abc-07f387e5f975" src="https://user-images.githubusercontent.com/103002884/211127546-70d14df4-7a88-408b-8573-8eb36e9b7ab1.png">

sfc-gh-tteixeira committed:
Small refactors.

sfc-gh-tteixeira commented:
Hey, @vdonato, maybe you can help me with this...
  
I can't run the e2e tests locally on my Mac, so update the screenshots. I'm getting this error:

```
circleci@947d527e5e8d:~/repo/frontend$ yarn cy:run --env updateSnapshots=true --spec ../e2e/specs/st_help.spec.js
yarn run v1.22.19
$ unset NODE_OPTIONS && cypress run --env updateSnapshots=true --spec ../e2e/specs/st_help.spec.js
It looks like this is your first time using Cypress: 6.9.1

  ✖  Verifying Cypress can run /home/circleci/.cache/Cypress/6.9.1/Cypress
    → Cypress Version: 6.9.1
Cypress failed to start.

This is usually caused by a missing library or dependency.

The error below should indicate which dependency is missing.

https://on.cypress.io/required-dependencies

If you are using Docker, we provide containers with all required dependencies installed.

----------

qemu: uncaught target signal 5 (Trace/breakpoint trap) - core dumped
qemu: uncaught target signal 5 (Trace/breakpoint trap) - core dumped
qemu: uncaught target signal 5 (Trace/breakpoint trap) - core dumped
qemu: uncaught target signal 5 (Trace/breakpoint trap) - core dumped
[396:0107/030403.391738:ERROR:bus.cc(393)] Failed to connect to the bus: Failed to connect to socket /run/dbus/system_bus_socket: No such file or directory
[396:0107/030403.432713:ERROR:file_path_watcher_linux.cc(315)] inotify_init() failed: Function not implemented (38)
[396:0107/030403.505420:WARNING:gpu_process_host.cc(1267)] The GPU process has crashed 1 time(s)
[396:0107/030403.516134:WARNING:gpu_process_host.cc(1267)] The GPU process has crashed 2 time(s)
[396:0107/030403.518726:WARNING:gpu_process_host.cc(1267)] The GPU process has crashed 3 time(s)
[396:0107/030403.521989:WARNING:gpu_process_host.cc(1267)] The GPU process has crashed 4 time(s)
[396:0107/030403.524232:WARNING:gpu_process_host.cc(1267)] The GPU process has crashed 5 time(s)
[396:0107/030403.526404:WARNING:gpu_process_host.cc(1267)] The GPU process has crashed 6 time(s)
[396:0107/030403.530306:WARNING:gpu_process_host.cc(1267)] The GPU process has crashed 7 time(s)
[396:0107/030403.561996:WARNING:gpu_process_host.cc(1267)] The GPU process has crashed 8 time(s)
[396:0107/030403.564703:WARNING:gpu_process_host.cc(1267)] The GPU process has crashed 9 time(s)
[396:0107/030403.564815:FATAL:gpu_data_manager_impl_private.cc(445)] GPU process isn't usable. Goodbye.
qemu: uncaught target signal 5 (Trace/breakpoint trap) - core dumped
[436:0107/030408.032688:ERROR:bus.cc(393)] Failed to connect to the bus: Failed to connect to socket /run/dbus/system_bus_socket: No such file or directory
[436:0107/030408.072780:ERROR:file_path_watcher_linux.cc(315)] inotify_init() failed: Function not implemented (38)
[436:0107/030408.134972:WARNING:gpu_process_host.cc(1267)] The GPU process has crashed 1 time(s)
[436:0107/030408.141924:WARNING:gpu_process_host.cc(1267)] The GPU process has crashed 2 time(s)
[436:0107/030408.144114:WARNING:gpu_process_host.cc(1267)] The GPU process has crashed 3 time(s)
[436:0107/030408.148424:WARNING:gpu_process_host.cc(1267)] The GPU process has crashed 4 time(s)
[436:0107/030408.201465:WARNING:gpu_process_host.cc(1267)] The GPU process has crashed 5 time(s)
[436:0107/030408.222011:WARNING:gpu_process_host.cc(1267)] The GPU process has crashed 6 time(s)
[436:0107/030408.222148:FATAL:gpu_data_manager_impl_private.cc(445)] GPU process isn't usable. Goodbye.
qemu: uncaught target signal 5 (Trace/breakpoint trap) - core dumped

----------

Platform: linux (Debian - 11.0)
Cypress Version: 6.9.1
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
circleci@947d527e5e8d:~/repo/frontend$ 
```

<img width="1007" alt="Screen Shot 2023-01-06 at 19 05 15" src="https://user-images.githubusercontent.com/103002884/211128747-59cdf796-2791-4bc3-8d5f-aac55d74ff00.png">

---

I'm running this in a Docker container with `$ ./scripts/run_e2e_tests.py -u e2e/specs/<name_of_test>.spec.js` as described [here](https://github.com/streamlit/streamlit/wiki/Running-e2e-tests-and-updating-snapshots#running-e2e-tests-using-docker).

The logs above were captured by running `cy:run` directly inside of the container, just so I get pretty colors. But the error when I call `run_e2e_tests.py` is the exact same.

I tried updating the deps as mentioned at https://on.cypress.io/required-dependencies, but that doesn't do it either. Any idea what's up? Could this be an ARM thing?

sfc-gh-tteixeira committed:
Support for Python 3.7

sfc-gh-tteixeira committed:
Add Github action to update snapshots (still in testing)

sfc-gh-tteixeira committed:
Store Cypress videos

sfc-gh-tteixeira committed:
Remove old screenshots

vdonato commented:
@sfc-gh-tteixeira whoops, sorry! The @mention from a few weeks back must have gotten lost in a sea of emails, so I didn't notice it until a later email notification for this PR.

I haven't updated snapshots locally via the Docker image for quite awhile at this point (I've mostly been deleting the snapshots to allow cypress running in CI to regenerate them for me), although I was under the impression that doing so still worked given that a few changes to the Dockerfile have been made relatively recently (CC @sfc-gh-kbregula and @sfc-gh-tszerszen as I believe they recently made some related changes). I'm not sure if folks still using the local docker test env are using M1 machines, though.

In either case, I think the path being taken in #6006 would likely be the preferred improvement here, anyway, given that attempting to replicate the exact environment used in CI to take snapshots locally has always been extremely fragile / oftentimes slower than using CI when updating tons of snapshots due to the lack of parallelization.

sfc-gh-kbregula commented:
I'm using Docker image on Mac OS x86, but building the image currently doesn't work because there is a problem with deleted options in `pip` tool.

To build the dockerfile, you need to add one patch locally.
```diff
diff --git a/Makefile b/Makefile
index 5b39d4606..a10b9bcb4 100644
--- a/Makefile
+++ b/Makefile
@@ -83,9 +83,9 @@ SHOULD_INSTALL_TENSORFLOW := $(shell python scripts/should_install_tensorflow.py
 py-test-install: lib/test-requirements.txt
        # As of Python 3.9, we're using pip's legacy-resolver when installing
        # test-requirements.txt, because otherwise pip takes literal hours to finish.
-       pip install -r lib/test-requirements.txt --use-deprecated=legacy-resolver
+       pip install -r lib/test-requirements.txt
 ifeq (${SHOULD_INSTALL_TENSORFLOW},true)
-       pip install -r lib/test-requirements-with-tensorflow.txt --use-deprecated=legacy-resolver
+       pip install -r lib/test-requirements-with-tensorflow.txt
 else
        @echo ""
        @echo "Your system does not support the official, pre-built tensorflow binaries."
```
And then follow the instructions:
https://github.com/streamlit/streamlit/wiki/Running-e2e-tests-and-updating-snapshots

tl;dr;
```bash
# Build or rebuild the test env image. This normally takes a long time.
$ make build-test-env
# The -u flag tells Cypress to update snapshots.
$ ./scripts/run_e2e_tests.py -u e2e/specs/<name_of_test>.spec.js
```

sfc-gh-tteixeira committed:
Fix issues mayagbarnes brought up.

sfc-gh-tteixeira committed:
Get test name from `github.events` in workflow-level statement.

sfc-gh-tteixeira committed:
Fix e2e test.

sfc-gh-tteixeira committed:
Merge remote-tracking branch 'origin/develop' into st-help

sfc-gh-tteixeira committed:
Readd import I removed by accident

sfc-gh-tteixeira committed:
Merge remote-tracking branch 'origin/develop' into st-help

sfc-gh-tteixeira committed:
Fix tests.

sfc-gh-tteixeira committed:
Add offset to screenshots.

sfc-gh-tteixeira committed:
Add snapshots

sfc-gh-tteixeira committed:
Fix Python 3.7 test

sfc-gh-tteixeira committed:
Test shouldn't depend on internal module for no reason.

sfc-gh-tteixeira committed:
Remove snapshots so we can rebuild them.

sfc-gh-tteixeira committed:
Update screenshots

sfc-gh-tteixeira commented:
Ok, I finally I figured this out. I had a bug in my e2e test, but Cypress was not telling me what it was. So I upgraded Cypress and the new version told me about the bug.

(Then I downgraded Cypress again, since otherwise this PR would get too long. I'll create a separate PR with the new Cypress later!)

Ready for a new review @vdonato

sfc-gh-tteixeira committed:
Remove screenshots that broke when I fixed the header issue. Need
update.

sfc-gh-tteixeira committed:
Fix magic e2e test.

sfc-gh-tteixeira committed:
Add snapshots for e2e tests that broke when I fixed the header isssue.

sfc-gh-tteixeira committed:
Add support for st.help() without args

sfc-gh-tteixeira committed:
Tiny CSS change to show a little bit of overflowing docs.

sfc-gh-tteixeira committed:
Add test.

sfc-gh-tteixeira commented on lib/streamlit/elements/doc_string.py:
```diff
@@ -31,26 +40,23 @@
+            The object whose information should be displayed. If `None`, this
+            call will display help for Streamlit itself.
 
         Example
```
@snehankekre I'd like to embed apps below each of these examples, to show the output of `st.help()` in each case. How do we do that nowadays?

snehankekre replied:
@sfc-gh-tteixeira The process is described in [How to embed apps in our docs](https://www.notion.so/How-to-embed-apps-in-our-docs-ea6b286e759a45caaf5769a3d0184ad2). According to the process, if the behavior of the app to be embedded depends on some features in this PR branch (i.e. not in `streamlit-nightly`), we should first merge this feature PR. After merging, we wait for the feature to be available in nightly and then follow the process described in that doc. 

The embedded apps will have to come as a follow up PR. I can take of adding them and will mark you as a reviewer on that PR.

snehankekre committed:
Enclose inline code in reStructuredText within two backticks

snehankekre commented:
Noticed some overflowing text when I tried this example with columns:

![image](https://user-images.githubusercontent.com/20672874/215709992-d2ad9a8f-07df-40ad-bdfe-5feab7cfb420.png)

```python
import pandas

import streamlit as st

col1, col2 = st.columns(2)

with col1:
    # Get help for Pandas DataFrame:
    pandas.DataFrame

with col2:
    # Get help for Streamlit itself:
    st
```

snehankekre commented on frontend/src/components/elements/DocString/styled-components.ts:
```diff
@@ -16,35 +16,86 @@
   })
 )
 
 export const StyledDocHeader = styled.div(({ theme }) => ({
```
To prevent overflow of text described in https://github.com/streamlit/streamlit/pull/5857#issuecomment-1409965353, there should either be this:

```suggestion
export const StyledDocHeader = styled.div(({ theme }) => ({
wordBreak: "break-word",
```

OR 

```suggestion
export const StyledDocHeader = styled.div(({ theme }) => ({
overflow: ["auto", "overlay"],
```

vdonato commented on lib/tests/streamlit/write_test.py:
```diff
@@ -181,6 +181,7 @@ def test_session_state(self):
 
     def test_snowpark_dataframe_write(self):
         """Test st.write with snowflake.snowpark.dataframe.DataFrame."""
+        from tests.streamlit.snowpark_mocks import DataFrame, Row
```
Is this import needed? Seems weird to me since the import already exists at the top of the file, so it'd be worth leaving a comment as to why in case there's some subtle reason this is required

sfc-gh-tteixeira replied:
I moved those imports inside the test so the test suite still runs even if you don't have that package. I'll add a comment about it.

(And then there was a silent merge conflict that put the imports back at the top of the file! Fixing now.)

vdonato commented on lib/tests/streamlit/write_test.py:
```diff
@@ -201,6 +202,7 @@ def test_snowpark_dataframe_write(self):
     )
     def test_pyspark_dataframe_write(self):
         """Test st.write with pyspark.sql.DataFrame."""
+        from tests.streamlit import pyspark_mocks
```
Same comment about imports here

vdonato commented on lib/tests/streamlit/elements/help_test.py:
```diff
@@ -46,143 +73,113 @@ def test_basic_func_without_doc(self):
-                ),
-            )
-        self.assertTrue(ds.doc_string.startswith("Function decorator to"))
+    # This doesn't even compile when running in 3.7, so I'm commenting it out.
+    # Which means we can't test support for walrus in st.help :(
```
nit: I'd add a TODO to support this test once we're able to.

I think we can technically drop support for 3.7 soon now that we have official support for 3.11, but I doubt we'll do so until we're forced to

vdonato commented on lib/tests/streamlit/elements/help_test.py:
```diff
@@ -193,23 +190,171 @@ def test_doc_defined_for_type(self):
+        ds = self.get_delta_from_queue().new_element.doc_string
+        self.assertEqual(len(ds.members), 5)
+
+        self.assertEqual(ds.members[0].name, "a")
+        self.assertEqual(ds.members[0].value, "1")
+        self.assertEqual(ds.members[0].doc_string, "")
+        self.assertEqual(ds.members[0].type, "int")
+
+        self.assertEqual(ds.members[1].name, "b")
+        self.assertEqual(ds.members[1].value, "2")
+        self.assertEqual(ds.members[1].doc_string, "")
+        self.assertEqual(ds.members[1].type, "int")
+
+        self.assertEqual(ds.members[2].name, "e")
+        self.assertEqual(ds.members[2].value, "")
+        self.assertEqual(ds.members[2].doc_string, "Property e")
+        self.assertEqual(ds.members[2].type, "property")
+
+        self.assertEqual(ds.members[3].name, "classmethod1")
+        self.assertEqual(ds.members[3].value, "")
+        self.assertEqual(ds.members[3].doc_string, "Class method 1")
+        self.assertEqual(ds.members[3].type, "method")
+
+        self.assertEqual(ds.members[4].name, "staticmethod1")
+        self.assertEqual(ds.members[4].value, "")
+        self.assertEqual(ds.members[4].doc_string, "Static method 1")
+        self.assertEqual(ds.members[4].type, "function")
```
nit: this could be made more concise by writing something like

```python
expected_members = [
    { 'name': 'a', 'value': '1', 'doc_string': '', 'type': 'int' },
    # ...
]

for i in range(len(expected_members)):
    self.assertEqual(ds.members[i].name, expected_members['name'])
    # ...
```

vdonato commented on lib/tests/streamlit/elements/help_test.py:
```diff
@@ -193,23 +190,171 @@ def test_doc_defined_for_type(self):
+        ds = self.get_delta_from_queue().new_element.doc_string
+        self.assertEqual(len(ds.members), 7)
+
+        self.assertEqual(ds.members[0].name, "a")
+        self.assertEqual(ds.members[0].value, "1")
+        self.assertEqual(ds.members[0].doc_string, "")
+        self.assertEqual(ds.members[0].type, "int")
+
+        self.assertEqual(ds.members[1].name, "b")
+        self.assertEqual(ds.members[1].value, "2")
+        self.assertEqual(ds.members[1].doc_string, "")
+        self.assertEqual(ds.members[1].type, "int")
+
+        self.assertEqual(ds.members[2].name, "c")
+        self.assertEqual(ds.members[2].value, "3")
+        self.assertEqual(ds.members[2].doc_string, "")
+        self.assertEqual(ds.members[2].type, "int")
+
+        self.assertEqual(ds.members[3].name, "d")
+        self.assertEqual(ds.members[3].value, "4")
+        self.assertEqual(ds.members[3].doc_string, "")
+        self.assertEqual(ds.members[3].type, "int")
+
+        self.assertEqual(ds.members[4].name, "e")
+        self.assertEqual(ds.members[4].value, "")
+        self.assertEqual(ds.members[4].doc_string, "Property e")
+        self.assertEqual(ds.members[4].type, "property")
+
+        self.assertEqual(ds.members[5].name, "classmethod1")
+        self.assertEqual(ds.members[5].value, "")
+        self.assertEqual(ds.members[5].doc_string, "Class method 1")
+        self.assertEqual(ds.members[5].type, "method")
+
+        self.assertEqual(ds.members[6].name, "staticmethod1")
+        self.assertEqual(ds.members[6].value, "")
+        self.assertEqual(ds.members[6].doc_string, "Static method 1")
+        self.assertEqual(ds.members[6].type, "function")
```
nit: same comment about cleaning this up to be more concise as above

vdonato commented on lib/tests/streamlit/elements/help_test.py:
```diff
@@ -193,23 +190,171 @@ def test_doc_defined_for_type(self):
+        self.assertEqual("MyClass", ds.name)
+        self.assertEqual(
+            "tests.streamlit.elements.help_test.StHelpTest."
+            "test_passing_a_class.<locals>.MyClass()",
```
Should the string here have parens at the end?

vdonato commented on lib/tests/streamlit/elements/help_test.py:
```diff
@@ -193,23 +190,171 @@ def test_doc_defined_for_type(self):
             pass
 
-        st.help(MyClass)
+        with patch_varname_getter():
+            st.help(MyClass)
```
Should this class be instantiated?

vdonato commented on lib/streamlit/elements/doc_string.py:
```diff
@@ -86,66 +125,416 @@ def _marshall(doc_string_proto: DocStringProto, obj: Any) -> None:
 
-    doc_string = inspect.getdoc(obj)
+    obj_value = _get_value(obj, var_name)
 
```
nit: can drop this newline

vdonato commented on lib/streamlit/elements/doc_string.py:
```diff
@@ -86,66 +125,416 @@ def _marshall(doc_string_proto: DocStringProto, obj: Any) -> None:
+    #   st.help(bar.Baz(123))
+    #
+    #   The name is Baz
+    name = getattr(obj, "__name__", None)
+    return name
```
nit: can just write

```suggestion
    return getattr(obj, "__name__", None)
```

vdonato commented on lib/streamlit/elements/doc_string.py:
```diff
@@ -86,66 +125,416 @@ def _marshall(doc_string_proto: DocStringProto, obj: Any) -> None:
+    scriptrunner_frame = _get_scriptrunner_frame()
+
+    if scriptrunner_frame is None:
+        # If there's no ScriptRunner frame, something weird is going on. This should never happen
+        # but let's bail out nicely just in case there's some valid edge case where this is OK.
```
I think one edge case where this may happen is if a user calls a script with `st.*` calls directly via `python my_script.py`, in which case we want every `st.*` call to be more or less a no-op and just want to print a message to the user telling them to use `streamlit run`

May be worth  mentioning this as the one place we expect this in the comment

sfc-gh-tteixeira replied:
Great catch. Will add.

vdonato commented on lib/streamlit/elements/doc_string.py:
```diff
@@ -86,66 +125,416 @@ def _marshall(doc_string_proto: DocStringProto, obj: Any) -> None:
+
+        members_for_sorting.append((weight, member))
+
+    if members_for_sorting:
+        _, members = zip(*sorted(members_for_sorting, key=lambda x: (x[0], x[1].name)))
+        return members
+
+    return []
```
nit: I think this can be rewritten to something a bit more readable to those unfamiliar with the `zip` function and splat operator by writing something like

```python3
sorted_members = sorted(members_for_sorting, key=lambda x: (x[0], x[1].name))
return [m for _, m in sorted_members]
```

sfc-gh-tteixeira replied:
Great point. Love it!

vdonato commented on lib/streamlit/elements/doc_string.py:
```diff
@@ -86,66 +125,416 @@ def _marshall(doc_string_proto: DocStringProto, obj: Any) -> None:
     return sig
+
+
+class UNSET:
+    pass
```
nit: This seems unused

vdonato reviewed (approved):
Overall this LGTM. Just a few things

* I left a few nits that aren't particularly important
* It'd likely be still worth getting someone to review the CSS 😅 
* Most of the helper functions added to `lib/streamlit/elements/doc_string.py` likely don't need their own individual unit tests because they're internal + the code paths are exercised thoroughly by tests in `help_test.py` and e2e tests, but I think it may be worthwhile to add a few tests for `_get_variable_name` given the somewhat complex AST parsing that happens in it

Also heads up that I'm going to be on vacation for a bit over a week starting Thursday 🌴! So if this needs more attention, it may be worth getting another member of the OS team to push it forward. I don't think there's much work that needs to be done before I'd be comfortable checking this into `develop`, though.

sfc-gh-tteixeira committed:
Improve legibility and comments.

sfc-gh-tteixeira committed:
Comments

sfc-gh-tteixeira committed:
Prettify tests.

sfc-gh-tteixeira committed:
Only return first line of multiline, and add more tests

sfc-gh-tteixeira committed:
Merge remote-tracking branch 'personal/st-help' into st-help

sfc-gh-tteixeira committed:
Merge branch 'develop' into st-help

sfc-gh-tteixeira committed:
Fix st.help header overflow when narrow.

sfc-gh-tteixeira commented:
> Noticed some overflowing text when I tried this example with columns:

Fixed!

> but I think it may be worthwhile to add a few tests for _get_variable_name

Added!

Ready for another review.

sfc-gh-tteixeira committed:
Update snapshots

sfc-gh-tteixeira commented:
The only failing test right now is `Python Versions / py_version (max)`, and it's failing because of this message in `python e2e/scripts/st_arrow_dataframe_column_types.py`:

```
2023-02-11 11:35:47.724 INFO    root: Applying automatic fixes for column types to make the dataframe Arrow-compatible.
Traceback (most recent call last):
  File "/Users/tteixeira/Projects/Streamlit/streamlit/lib/streamlit/type_util.py", line 759, in data_frame_to_bytes
    table = pa.Table.from_pandas(df)
  File "pyarrow/table.pxi", line 3557, in pyarrow.lib.Table.from_pandas
  File "/Users/tteixeira/.local/share/virtualenvs/lib-UsUdzUK_/lib/python3.10/site-packages/pyarrow/pandas_compat.py", line 611, in dataframe_to_arrays
    arrays = [convert_column(c, f)
  File "/Users/tteixeira/.local/share/virtualenvs/lib-UsUdzUK_/lib/python3.10/site-packages/pyarrow/pandas_compat.py", line 611, in <listcomp>
    arrays = [convert_column(c, f)
  File "/Users/tteixeira/.local/share/virtualenvs/lib-UsUdzUK_/lib/python3.10/site-packages/pyarrow/pandas_compat.py", line 598, in convert_column
    raise e
  File "/Users/tteixeira/.local/share/virtualenvs/lib-UsUdzUK_/lib/python3.10/site-packages/pyarrow/pandas_compat.py", line 592, in convert_column
    result = pa.array(col, type=type_, from_pandas=True, safe=safe)
  File "pyarrow/array.pxi", line 316, in pyarrow.lib.array
  File "pyarrow/array.pxi", line 79, in pyarrow.lib._ndarray_to_array
  File "pyarrow/array.pxi", line 67, in pyarrow.lib._ndarray_to_type
  File "pyarrow/error.pxi", line 121, in pyarrow.lib.check_status
pyarrow.lib.ArrowNotImplementedError: ('Unsupported numpy type 15', 'Conversion failed for column complex with type complex128')
```

Since I didn't touch anything related to that, I cloned the `streamlit/streamlit` repo in a separate folder, ran the test above on `develop`, and found the same issue. So this appears to be a problem with the main branch.

My guess is this issue was introduced in https://github.com/streamlit/streamlit/pull/6022, but, then again, I'm not sure why tests have been passing since that PR got merged. Any idea what's up, @vdonato?

lukasmasuch commented on lib/tests/streamlit/elements/help_test.py:
```diff
@@ -193,23 +194,270 @@ def test_doc_defined_for_type(self):
+
+    def test_if_dont_know_just_echo(self):
+        tests = [
+            (test := "foo()", test),
```
This syntax does not seem to be supported in Python 3.7 :( The python min tests are falling because of invalid syntax here: https://github.com/streamlit/streamlit/actions/runs/4150469302/jobs/7180147008#step:11:52

<img width="928" alt="image" src="https://user-images.githubusercontent.com/2852129/218543103-245c26f8-5997-4a59-a992-1b10b0b2891e.png">

willhuang1997 replied:
Huh. How does that happen? Shouldn't that be found in the PR request?

lukasmasuch replied:
> Shouldn't that be found in the PR request?

You probably are referring to pre-commit or? Pre-commit only checks based on the installed python version, which is most likely > 3.7

sfc-gh-tteixeira replied:
Good find. Thanks for the help debugging!

lukasmasuch commented:
The `ArrowNotImplementedError` is expected to be printed out in this e2e tests, but should not cause the tests to fail. The python min tests seem to be falling because of unsupported syntax (see comment above). But I'm not sure yet why `max` is falling 🤔 It stops at `python /home/runner/work/streamlit/streamlit/e2e/scripts/st_help.py`, but there isn't any additional error information: 

<img width="698" alt="image" src="https://user-images.githubusercontent.com/2852129/218544177-8f08e866-fd34-40b9-b5d9-b5fe5b058331.png">

I suspect that the script somehow exits because it is executed via `python` and not via `streamlit`.. but that's just a hypothesis.

lukasmasuch commented:
If I execute the `st_help` e2e script locally via python, I get the following error: 

```
❯ python e2e/scripts/st_help.py                                      
2023-02-13 19:53:43.576 WARNING streamlit.config: 
Warning: the config option 'server.enableCORS=false' is not compatible with 'server.enableXsrfProtection=true'.
As a result, 'server.enableCORS' is being overridden to 'true'.

More information:
In order to protect against CSRF attacks, we send a cookie with each request.
To do so, we must specify allowable origins, which places a restriction on
cross-origin resource sharing.

If cross origin resource sharing is required, please disable server.enableXsrfProtection.
            
2023-02-13 19:53:43.577 WARNING root: 
  Warning: to view this Streamlit app on a browser, run it with the following
  command:

    streamlit run e2e/scripts/st_help.py [ARGUMENTS]
Traceback (most recent call last):
  File "/Users/Shared/Workspace/Development/streamlit/e2e/scripts/st_help.py", line 31, in <module>
    st.help(FooWithNoDocs)
  File "/Users/Shared/Workspace/Development/streamlit/lib/streamlit/runtime/metrics_util.py", line 311, in wrapped_func
    result = non_optional_func(*args, **kwargs)
  File "/Users/Shared/Workspace/Development/streamlit/lib/streamlit/elements/doc_string.py", line 114, in help
    _marshall(doc_string_proto, obj)
  File "/Users/Shared/Workspace/Development/streamlit/lib/streamlit/elements/doc_string.py", line 128, in _marshall
    var_name = _get_variable_name()
  File "/Users/Shared/Workspace/Development/streamlit/lib/streamlit/elements/doc_string.py", line 235, in _get_variable_name
    return _get_variable_name_from_code_str(code)
  File "/Users/Shared/Workspace/Development/streamlit/lib/streamlit/elements/doc_string.py", line 239, in _get_variable_name_from_code_str
    tree = ast.parse(code)
  File "/opt/homebrew/Cellar/python@3.10/3.10.9/Frameworks/Python.framework/Versions/3.10/lib/python3.10/ast.py", line 50, in parse
    return compile(source, filename, mode, flags,
TypeError: compile() arg 1 must be a string, bytes or AST object
```

I assume this is what is causing the Python Max to fail, but not sure why the error is not printed out in the CI logs.

mayagbarnes commented on frontend/src/components/elements/DocString/styled-components.ts:
```diff
@@ -16,35 +16,87 @@
+
+export const StyledDocValue = styled.span()
+
 export interface StyledDocContainerProps {
   width: number
 }
 
 export const StyledDocContainer = styled.span<StyledDocContainerProps>(
   ({ theme, width }) => ({
```
Nit: Looks like the `width` prop is unused now and can be removed

mayagbarnes commented on frontend/src/components/elements/DocString/styled-components.ts:
```diff
@@ -16,35 +16,87 @@
+
+  overflow: ["auto", "overlay"],
+
+  "&:not(:last-child)": {
+    borderBottom: `1px solid ${theme.colors.fadedText05}`,
+  },
```
Just curious why bottom border is implemented with the not the last child selector. Probably missing something but playing with this in inspect and can't discern the difference between using this selector and directly specifying the bottom border.

sfc-gh-tteixeira replied:
Oh, I think this is left over from a previous version of the st.help spec I was working with. I'll remove it.

In a previous version of the spec, it was possible for this header element to be the last child of the `DocString` element, and this selector avoided a double bottom border. (This would happen when the object passed to `st.help( <-- here --> )` doesn't have a docstring or member variables/methods.)

However, since we now show a placeholder saying "No docs available" when there's no docstring, this header can never be the last child anymore.

mayagbarnes reviewed (approved):
LGTM from CSS perspective 👍🏼

vdonato commented:
@sfc-gh-tteixeira: I had some time to poke around at things, and I think that @LukasMasuch's evaluation is correct

* `py_version (min)` tests are failing because one test in `tests/streamlit/elements/help_test.py` was written using the walrus operator so blows up since our current min version is 3.7
* `py_version (max)` tests are failing because running `python e2e/scripts/st_help.py` explodes when it hits the `st.help(FooWithNoDocs)` line. This is because the `_get_variable_name_from_code_str()` helper function call fails in a "raw script run" context since no script runner exists: `_get_current_line_of_code_as_str()` returns `None`, which blows up when it's passed as an argument to `ast.parse()`.
   * I think this is solvable by just returning the empty string when `None` is passed to `_get_variable_name_from_code_str()`
* As far as I can tell, while running `python e2e/scripts/st_arrow_dataframe_column_types.py` does print an error to the console, the error code returned by running that command is still 0, so it shouldn't cause the make target to fail

vdonato commented on lib/streamlit/elements/doc_string.py:
```diff
@@ -234,28 +231,12 @@ def _get_variable_name():
-        # example. If this happens, we can't determine the variable name. Just return.
-        # For the background on why exec() doesn't produce code_context, see
-        # https://stackoverflow.com/a/12072941
-        return None
```
Ah, I'm guessing this is the small refactor that broke things when running a script using `python3 <script_name>` instead of `streamlit run <script_name>` while keeping the behavior the same when running within streamlit

vdonato reviewed (approved):
LGTM once the last couple test issues are resolved

snehankekre reviewed (approved):
Docstring changes LGTM :shipit:

sfc-gh-tteixeira committed:
Merge remote-tracking branch 'origin/develop' into st-help

sfc-gh-tteixeira committed:
Make st_help test work in 3.7

sfc-gh-tteixeira committed:
Remove unecessary stuff.

sfc-gh-tteixeira committed:
Fix pure Python run of st_help.py

sfc-gh-tteixeira committed:
Differentiate st.help() from st.help(None)

sfc-gh-tteixeira commented:
@vdonato @LukasMasuch : you're totally right. The Arrow thing was a red herring! It's just a warning 🤦 

Fixed now.

Thanks for the help debugging!

sfc-gh-tteixeira commented:
Markes sense! LGTM then.

sfc-gh-tteixeira committed:
Fix typo in docstring

sfc-gh-tteixeira commented:
@jrieke here's a demo:
https://streamlit-feature-demos-st-helpstreamlit-app-125gy2.streamlit.app/

jrieke commented:
Demo app looks great! Can you write QA plan and have this QAed whenever you're ready? I'm OOO until Friday, so you can skip my approval and just send it off to Kasim directly :)

vdonato commented on proto/streamlit/proto/DocString.proto:
```diff
@@ -18,19 +18,35 @@ syntax = "proto3";
 
 // Formatted text
 message DocString {
-
-  // The name of the object.
+  // The name the user gave to the variable holding this object.
   string name = 1;
```
Whoops, my bad for not noticing this earlier, especially given that I have code ownership over these files specifically to enforce this 😅 

We'll need to ensure that these proto changes are backwards compatible given that other stakeholders rely on them

We also want to be a bit more strict than normal proto backwards compatibility guidelines to also avoid changing field names to make things easier on folks when they're upgrading the OS lib version.

sfc-gh-tteixeira replied:
Done. PTAL.

sfc-gh-tteixeira committed:
Properly upgrade DocString.proto

sfc-gh-tteixeira committed:
Merge remote-tracking branch 'origin/develop' into st-help

sfc-gh-tteixeira committed:
Move "reserved" line near the previous location.

vdonato commented on proto/streamlit/proto/DocString.proto:
```diff
@@ -19,18 +19,47 @@ syntax = "proto3";
+  //string signature = 5;
+  reserved 5;
+
+  // The name the user gave to the variable holding this object.
+  string name = 6;
```
nit: I think we could probably just keep the current name field vs adding this one, but not a big deal either way

sfc-gh-tteixeira replied:
IMO the meaning changed enough that the right thing to do is to deprecate.

vdonato reviewed (approved):
new proto LGTM

sfc-gh-tteixeira committed:
Merge remote-tracking branch 'origin/develop' into st-help

sfc-gh-tteixeira committed:
Fix 3.7 test

sfc-gh-tteixeira committed:
Merge remote-tracking branch 'origin/develop' into st-help

sfc-gh-tteixeira committed:
Fix tests

sfc-gh-tteixeira committed:
Fix tests

sfc-gh-tteixeira committed:
Fix off-by-one bug on 3.7 due to :-1 rather than :None

sfc-gh-tteixeira committed:
Ok, final fix for tests???

sfc-gh-tteixeira committed:
Merge remote-tracking branch 'origin/develop' into st-help
